cnd/src/http_api/routes.rs

```diff
@@ -33,88 +34,105 @@ pub fn into_rejection(problem: HttpApiProblem) -> Rejection {
-
-            return Ok(empty_swap);
-        }
-    };
-
-    match finalized_swap.role {
-        Role::Alice => {
-            let state = AliceHanEthereumHalightBitcoinState {
-                alpha_ledger_state,
-                beta_ledger_state,
-                finalized_swap,
-            };
-
-            let maybe_action_names = vec![
-                state.init_action().map(|_| "init"),
-                state.fund_action().map(|_| "fund"),
-                state.redeem_action().map(|_| "redeem"),
-                state.refund_action().map(|_| "refund"),
-            ];
-            make_finalized_swap_entity(swap_id, state, maybe_action_names)
-        }
-        Role::Bob => {
-            let state = BobHanEthereumHalightBitcoinState {
-                alpha_ledger_state,
-                beta_ledger_state,
-                finalized_swap,
-            };
+) -> anyhow::Result<siren::Entity>
+where
+    Facade: Load<InProgressSwap<han::InProgressSwap, halight::InProgressSwap>>
+        + Load<CreatedSwap<han::CreatedSwap, halight::CreatedSwap>>,
+{
+    let alpha_ledger_state: LedgerState<
+        asset::Ether,
+        htlc_location::Ethereum,
+        transaction::Ethereum,
+    > = facade
+        .alpha_ledger_states
+        .get(&swap_id)
+        .await?
+        .ok_or_else(|| {
+            tracing::error!("failed to find alpha ledger state");
+            HttpApiProblem::with_title_and_type_from_status(StatusCode::INTERNAL_SERVER_ERROR)
+        })?;
 
-            // Bob cannot init and refund in this swap combination
-            let maybe_action_names = vec![
-                state.fund_action().map(|_| "fund"),
-                state.redeem_action().map(|_| "redeem"),
-            ];
-            make_finalized_swap_entity(swap_id, state, maybe_action_names)
+    let beta_ledger_state = facade
+        .beta_ledger_states
+        .get(&swap_id)
+        .await?
+        .ok_or_else(|| {
+            tracing::error!("failed to find beta ledger state");
+            HttpApiProblem::with_title_and_type_from_status(StatusCode::INTERNAL_SERVER_ERROR)
+        })?;
+
+    let created_swap =
+        match Load::<CreatedSwap<han::CreatedSwap, halight::CreatedSwap>>::load(&facade, swap_id)
+            .await?
+        {
+            Some(swap) => swap,
+            None => {
+                let empty_swap = siren::Entity::default().with_class_member("swaps");
+                tracing::debug!(
+                    "returning empty siren document because states are not yet completed"
+                );
+                return Ok(empty_swap);
+            }
+        };
+
+    let swap = Load::<InProgressSwap<han::InProgressSwap, halight::InProgressSwap>>::load(
+        &facade, swap_id,
+    )
+    .await?;
+    match swap {
+        None => make_created_swap_entity(swap_id, HanEthereumHalightBitcoinCreatedState {
+            created_swap: created_swap.into(),
+        }),
```

Comment:
Why do we have to go to the database twice here? Could we not just have an API that returns us exactly what we need (taking care of the difference between `InProgressSwap` and `CreatedSwap` inside of the datastructure by having some things be `Option`?

This looks to me like we are unnecessarily forcing this part of the codebase to work with the separation of `Created` and `InProgress` swaps and hence yielding sub-optimal code because we have to go to the database twice.

We don't need to find a model that works for every usecase in the codebase, we can have different models for different purposes :)
At this stage, we probably also want to re-consider, which fields the API actually needs and not load those (not for performance reasons, but to make the different needs more obvious and to avoid someone refactoring away "duplication" when there is none)